src/DnsClient/DnsTcpMessageHandler.cs

```diff
@@ -306,20 +339,54 @@ public ClientEntry GetNextClient()
                 {
                     while (entry == null && !TryDequeue(out entry))
                     {
-                        entry = new ClientEntry(new TcpClient(_endpoint.AddressFamily) { LingerState = new LingerOption(true, 0) }, _endpoint);
-                        entry.Client.Connect(_endpoint.Address, _endpoint.Port);
+                        entry = ConnectNew();
                     }
                 }
                 else
                 {
-                    entry = new ClientEntry(new TcpClient(_endpoint.AddressFamily), _endpoint);
-                    entry.Client.Connect(_endpoint.Address, _endpoint.Port);
+                    entry = ConnectNew();
                 }
 
                 return entry;
+
+                ClientEntry ConnectNew()
+                {
+                    var newClient = new TcpClient(_endpoint.AddressFamily)
+                    {
+                        LingerState = new LingerOption(true, 0)
+                    };
+
+                    bool gotCanceled = false;
+                    cancellationToken.Register(() =>
+                    {
+                        gotCanceled = true;
+                        newClient.Dispose();
+                    });
+
+                    try
+                    {
+                        newClient.Connect(_endpoint.Address, _endpoint.Port);
+                    }
+                    catch (Exception) when (gotCanceled)
+                    {
+                        throw new TimeoutException("Connection timed out.");

```

Comment:
The caller `Query` method doesn't seem to translate cancellations to `TimeoutException`.

Reply:
Not exactly sure what you mean by that.
What I want to do here is allowing LookupClient does handle OperationCanceled- and Timeout-Exceptions and retry the request if possible.
For TCP, if the connection attempt takes longer then the configured timeout for a query, I want to handle it the same way and maybe have LookupClient retry.

Reply:
changed both, sync and async version to `throw new OperationCanceledException("Connection timed out.", cancellationToken);` 
Still, its mostly a signal for LookupClient to retry